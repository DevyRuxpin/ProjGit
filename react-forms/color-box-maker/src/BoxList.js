import React, { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';

function BoxList() {
  const [boxes, setBoxes] = useState([]);

  const addBox = (newBox) => {
    setBoxes(boxes => [...boxes, { ...newBox, id: Date.now() }]);
  };

  const removeBox = (id) => {
    setBoxes(boxes => boxes.filter(box => box.id !== id));
  };

  return (
    <div>
      <NewBoxForm addBox={addBox} />
      <div className="BoxList">
        {boxes.map(box => (
          <Box
            key={box.id}
            id={box.id}
            width={box.width}
            height={box.height}
            backgroundColor={box.backgroundColor}
            removeBox={removeBox}
          />
        ))}
      </div>
    </div>
  );
}

export default BoxList; 