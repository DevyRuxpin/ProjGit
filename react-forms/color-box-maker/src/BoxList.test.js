import { render, screen, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

test('renders without crashing', () => {
  render(<BoxList />);
});

test('matches snapshot', () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

test('renders form inputs', () => {
  render(<BoxList />);
  expect(screen.getByLabelText(/width/i)).toBeInTheDocument();
  expect(screen.getByLabelText(/height/i)).toBeInTheDocument();
  expect(screen.getByLabelText(/background color/i)).toBeInTheDocument();
  expect(screen.getByText(/add box/i)).toBeInTheDocument();
});

test('can add a new box', () => {
  render(<BoxList />);
  
  const widthInput = screen.getByLabelText(/width/i);
  const heightInput = screen.getByLabelText(/height/i);
  const colorInput = screen.getByLabelText(/background color/i);
  const submitButton = screen.getByText(/add box/i);

  fireEvent.change(widthInput, { target: { value: '100' } });
  fireEvent.change(heightInput, { target: { value: '100' } });
  fireEvent.change(colorInput, { target: { value: 'red' } });
  fireEvent.click(submitButton);

  // Check that the box was added (should have a remove button)
  expect(screen.getByText('X')).toBeInTheDocument();
});

test('can remove a box', () => {
  render(<BoxList />);
  
  // Add a box first
  const widthInput = screen.getByLabelText(/width/i);
  const heightInput = screen.getByLabelText(/height/i);
  const colorInput = screen.getByLabelText(/background color/i);
  const submitButton = screen.getByText(/add box/i);

  fireEvent.change(widthInput, { target: { value: '100' } });
  fireEvent.change(heightInput, { target: { value: '100' } });
  fireEvent.change(colorInput, { target: { value: 'red' } });
  fireEvent.click(submitButton);

  // Remove the box
  const removeButton = screen.getByText('X');
  fireEvent.click(removeButton);

  // Check that the box was removed
  expect(screen.queryByText('X')).not.toBeInTheDocument();
});

test('form clears after submission', () => {
  render(<BoxList />);
  
  const widthInput = screen.getByLabelText(/width/i);
  const heightInput = screen.getByLabelText(/height/i);
  const colorInput = screen.getByLabelText(/background color/i);
  const submitButton = screen.getByText(/add box/i);

  fireEvent.change(widthInput, { target: { value: '100' } });
  fireEvent.change(heightInput, { target: { value: '100' } });
  fireEvent.change(colorInput, { target: { value: 'red' } });
  fireEvent.click(submitButton);

  // Check that form inputs are cleared
  expect(widthInput.value).toBe('');
  expect(heightInput.value).toBe('');
  expect(colorInput.value).toBe('');
}); 