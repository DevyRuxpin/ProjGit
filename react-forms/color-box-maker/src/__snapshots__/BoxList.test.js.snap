// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <div>
        <label
          for="width"
        >
          Width:
        </label>
        <input
          id="width"
          name="width"
          required=""
          type="number"
          value=""
        />
      </div>
      <div>
        <label
          for="height"
        >
          Height:
        </label>
        <input
          id="height"
          name="height"
          required=""
          type="number"
          value=""
        />
      </div>
      <div>
        <label
          for="backgroundColor"
        >
          Background Color:
        </label>
        <input
          id="backgroundColor"
          name="backgroundColor"
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Add Box
      </button>
    </form>
    <div
      class="BoxList"
    />
  </div>
</DocumentFragment>
`;
